File: pset6/credit/credit.py
import math
from sys import exit


def main():
    number = ask_number()
    number = [int(x) for x in str(number)]
    # print(number)
    sum1 = 0
    # goes through numbers in the list starting from second to last
    for i in range(len(number)-2, -1, -2):
        dig = number[i] * 2
        if dig >= 10:
            dig = 1 + dig % 10
        sum1 += dig
    
    sum2 = 0
    # goes through numbers from last
    for i in range(len(number)-1, -1, -2):
        sum2 += number[i]
    
    if (sum1 + sum2) % 10 == 0:
        if number[0] == 4:
            print("VISA")
        elif number[0] == 3 and (number[1] == 4 or number[1] == 7):
            print("AMEX")
        elif number[0] == 5 and (number[1] >= 1 and number[1] <= 5):
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")
    

def ask_number():
    while True:
        # imputs number and calculates length
        try:
            number = int(input("Number: "))
            digits = int(math.log10(number))+1
        # if inputted not number, try again
        except ValueError:
            continue
        else:
            # if length is not ok, exit
            if number < 0 or digits < 12 or digits > 16:
                print("INVALID")
                exit(1)
            return number
            break


main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

long ask_number(void);
int sum1(long nr);
int sum2(long nr);
int starting(long nr);
int main(void)
{
    long nr = ask_number();
    if (nr < pow(10, 12) || nr > pow(10, 16))
    {
        printf("INVALID\n");
    }
    else
    {
        int multiplied = sum1(nr);
        int sum = sum2(nr);
        int total = multiplied + sum;
        int first = starting(nr);
        if (total % 10 == 0) //tikrina ar algoritmo paskutinis skaitmuo yra 0
        {
            if (first == 4) //VISA pirmas skaicius 4
            {
                printf("VISA\n");
            }
            else if (first == 34 || first == 37) //AMEX pirmi skaiciai
            {
                printf("AMEX\n");
            }
            else if (first >= 51 && first <= 55) //MASTERCARD pirmi skaiciai
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

long ask_number(void) //klausia koks nr, ir paduoda numeri
{res = [int(x) for x in str(num)]
    long i;
    i = get_long("Number: "); //takes a number
    return i;
}

int sum1(long nr) //skaiciuoja kas antro skaiciaus daugyba ir suma
{
    int sum = 0;
    for (int i = 2; i < 17; i = i + 2)
    {
        long a, b, c, d, e, x, y;
        x = pow(10, i);
        y = pow(10, i - 1);
        a = nr % x;
        b = nr % y;
        c = a - b;
        d = c / y;
        e = d * 2;
    
        if (e >= 10) //jeigu gaunasi dvizenklis skaicius, tai padaro is abieju skaitmenu suma
        {
            e = 1 + e % 10;
        }
//    printf("Chosen nummber sum1: %li\n", d); //prints chosen numbers
        sum = sum + e;
    }
//    printf("sum1%i\n", sum);
    return sum;
}

int sum2(long nr) //skaiciuoja kita suma kur be daugybos
{
    int sum = 0;
    for (int i = 1; i < 17; i = i + 2)
    {
        long a, b, c, d, x, y;
        x = pow(10, i);
        y = pow(10, i - 1);
        a = nr % x;
        b = nr % y;
        c = a - b;
        d = c / y;
//    printf("Chosen nummber sum2: %li\n", d); //prints chosen numbers
        sum = sum + d;
    }
//    printf("sum2%i\n", sum);
    return sum;
}

int starting(long nr) //skaiciuoja kokio ilgio skaicius ir paima pirmuosius skaicius
{
    int d;
    int i = 1;
    int length, first, second;
    long a, b, c, x, y;
    long j = 0;
    for (length = 1; nr > j; length++)
    {
        j = pow(10, length + 1);
    }
    x = pow(10, length - 0);
    y = pow(10, length - 1);
    a = nr % x;
    b = nr % y;
    c = a - b;
    first = c / y;
    if (first != 4)
    {
        x = pow(10, length - 1);
        y = pow(10, length - 2);
        a = nr % x;
        b = nr % y;
        c = a - b;
        first = first * 10 + c / y;
    }
    return first;
}